File: src/brs/crypto/EncryptedData.java
package brs.crypto;

import brs.SignumException;
import signumj.entity.EncryptedMessage;

import java.nio.ByteBuffer;

// TODO replace this class with the one from burstkit4j
public class EncryptedData {
  private static final EncryptedData EMPTY_DATA = new EncryptedData(new byte[0], new byte[0]);

  public static EncryptedData encrypt(byte[] plaintext, byte[] myPrivateKey, byte[] theirPublicKey) {
    if (plaintext.length == 0) {
      return EMPTY_DATA;
    }
    EncryptedMessage message = Crypto.signumCrypto.encryptBytesMessage(plaintext, myPrivateKey, theirPublicKey);
    return new EncryptedData(message.getData(), message.getNonce());
  }

  public static EncryptedData readEncryptedData(ByteBuffer buffer, int length, int maxLength)
    throws SignumException.NotValidException {
    if (length == 0) {
      return EMPTY_DATA;
    }
    if (length > maxLength) {
      throw new SignumException.NotValidException("Max encrypted data length exceeded: " + length);
    }
    byte[] noteBytes = new byte[length];
    buffer.get(noteBytes);
    byte[] noteNonceBytes = new byte[32];
    buffer.get(noteNonceBytes);
    return new EncryptedData(noteBytes, noteNonceBytes);
  }

  private final byte[] data;
  private final byte[] nonce;

  public EncryptedData(byte[] data, byte[] nonce) {
    this.data = data;
    this.nonce = nonce;
  }

  public byte[] decrypt(byte[] myPrivateKey, byte[] theirPublicKey) {
    if (data.length == 0) {
      return data;
    }
    return Crypto.signumCrypto.decryptMessage(new EncryptedMessage(data, nonce, false), myPrivateKey, theirPublicKey);
  }

  public byte[] getData() {
    return data;
  }

  public byte[] getNonce() {
    return nonce;
  }

  public int getSize() {
    return data.length + nonce.length;
  }
}

File: src/main/kotlin/brs/crypto/EncryptedData.kt
package brs.crypto

import brs.BurstException
import burst.kit.entity.BurstEncryptedMessage

import java.nio.ByteBuffer

// TODO replace this class with the one from burstkit4j
class EncryptedData(val data: ByteArray, val nonce: ByteArray) {

    val size: Int
        get() = data.size + nonce.size

    fun decrypt(myPrivateKey: ByteArray, theirPublicKey: ByteArray): ByteArray {
        return if (data.isEmpty()) {
            data
        } else burstCrypto.decryptMessage(BurstEncryptedMessage(data, nonce, false), myPrivateKey, theirPublicKey)
    }

    companion object {
        private val EMPTY_DATA = EncryptedData(ByteArray(0), ByteArray(0))

        fun encrypt(plaintext: ByteArray, myPrivateKey: ByteArray, theirPublicKey: ByteArray): EncryptedData {
            if (plaintext.isEmpty()) {
                return EMPTY_DATA
            }
            val message = burstCrypto.encryptBytesMessage(plaintext, myPrivateKey, theirPublicKey)
            return EncryptedData(message.data, message.nonce)
        }

        fun readEncryptedData(buffer: ByteBuffer, length: Int, maxLength: Int): EncryptedData {
            if (length == 0) {
                return EMPTY_DATA
            }
            if (length > maxLength) {
                throw BurstException.NotValidException("Max encrypted data length exceeded: $length")
            }
            val noteBytes = ByteArray(length)
            buffer.get(noteBytes)
            val noteNonceBytes = ByteArray(32)
            buffer.get(noteNonceBytes)
            return EncryptedData(noteBytes, noteNonceBytes)
        }
    }
}
